moveit_core/planning_interface/include/moveit/planning_interface/planning_interface.h

```diff
@@ -75,6 +75,9 @@ struct PlannerConfigurationSettings
 /** \brief Map from PlannerConfigurationSettings.name to PlannerConfigurationSettings */
 typedef std::map<std::string, PlannerConfigurationSettings> PlannerConfigurationMap;
 
+using StateCostFn = std::function<double(const moveit::core::RobotState& robot_state, const MotionPlanRequest& request,
```

Comment:
I agree, ideally the cost function signature should only use primitive types so that it's easy to adapt for the different planners (see STOMP cost functions). Also, costs should rather be part of the MotionPlanRequest and not depend on them. STOMP currently has a generic cost function for all trajectory waypoints (https://github.com/ros-planning/moveit2/blob/main/moveit_planners/stomp/include/stomp_moveit/stomp_moveit_task.hpp#LL65C83-L65C88). For a simple state cost, I would therefore propose something like: `CostFn = std::function<double(const Eigen::VectorXd& values, bool& validity)` where validity is even up for debate. The actual implementation and MoveIt-specific state of the cost functions can likewise be managed with lambda bindings. Is there an issue with supporting this type of function with OMPL?

Reply:
Also, consider the [StateValidatorFn](https://github.com/ros-planning/moveit2/blob/main/moveit_planners/stomp/include/stomp_moveit/cost_functions.hpp#L53) as a use case that could be applicable to all planners.

Reply:
I've updated the signature to `std::function<double(const std::vector<double>& state_vector)` & you're right it simplifies using the cost function a lot. I also like the idea of adding the cost function to the MotionPlanRequest since that will simplify the implementation even more :+1:

Reply:
I have to investigate if the concept of a state validator function is applicable to other planners. How does it differ from the processing of constraints within the planner?